omegaconf/base.py

```diff
@@ -413,14 +414,36 @@ def _resolve_interpolation_from_parse_tree(
                 raise
             return None
 
-        assert resolved is not None
-        if isinstance(resolved, str):
-            # Result is a string: create a new StringNode for it.
-            return StringNode(
-                value=resolved,
-                key=key,
+        # If the output is not a Node already (e.g., because it is the output of a
+        # custom resolver), then we will need to wrap it within a Node.
+        must_wrap = not isinstance(resolved, Node)
+
+        # If the node is typed, validate (and possibly convert) the result.
+        if isinstance(value, ValueNode) and not isinstance(value, AnyNode):
```

Comment:
I had actually tried to do some kind of validation based on `_validate_set()` but ran into some issues (can't remember the details), so I figured it could be left for future work.

That being said, while writing doc to explain that no such validation occurred for config nodes, I realized that it actually did. This is because I'm passing `value._metadata.ref_type` to `_node_wrap()`, which will do the validation / conversion. So instead of updating the doc I added tests for it in 4af9c8b.